packages/dxf-deserializer/translate.js

```diff
@@ -0,0 +1,880 @@
+        break
+      case 'layer':
+        console.log('##### layer')
+        current = translateCurrent(current, layers, parts, options)
+        parts = []
+        // save the layer for later reference
+        obj['objects'] = [] // with a list of objects
+        layers.push(obj)
+        break
+      case 'variable':
+        console.log(JSON.stringify(obj))
+        current = translateCurrent(current, layers, parts, options)
+        parts = []
+        saveVariable(obj, options)
+        break
+
+      // 3D entities
+      case '3dface':
+        console.log('##### 3dface')
+        p = instantiatePolygon(obj, layers, options)
+//console.log(JSON.stringify(p))
+        if (current === null) {
+          console.log('##### start of 3dfaces CSG')
+          current = {type: '3dfaces'}
+          current['name'] = 'jscad' + numobjs
+          numobjs = numobjs + 1
+        }
+        break
+      case 'mesh':
+// FIXME complete translation
+        console.log('##### mesh')
+        current = translateCurrent(current, layers, parts, options)
+        parts = []
+        //objects.push(instantiateMesh(obj, layers, options))
+        translateMesh(obj, layers, options)
+        break
+
+      // 2D or 3D entities
+      case 'arc':
+        console.log('##### arc')
+        current = translateCurrent(current, layers, parts, options)
+        parts = []
+        translateArc(obj, layers, options)
+        break
+      case 'circle':
+        console.log('##### circle')
+        current = translateCurrent(current, layers, parts, options)
+        translateCircle(obj, layers, options)
+        parts = []
+        break
+      case 'ellipse':
+        console.log('##### ellipse')
+        current = translateCurrent(current, layers, parts, options)
+        parts = []
+        translateEllipse(obj, layers, options)
+        break
+      case 'line':
+        console.log('##### line')
+        current = translateCurrent(current, layers, parts, options)
+        parts = []
+        translateLine(obj, layers, options)
+        break
+      case 'polyline':
+        if (current === null) {
+          console.log('##### start of polyline')
+          current = getPolyType(obj)
+          current['name'] = 'jscad' + numobjs
+          numobjs = numobjs + 1
+        }
+        break
+      case 'vertex':
+        //console.log('##### vertex')
+        p = instantiateVector(obj)
+        break
+      case 'seqend':
+        current = translateCurrent(current, layers, parts, options)
+        parts = []
+        break
+
+      // 2D entities
+      case 'lwpolyline':
+        console.log('##### lwpolyline')
+        current = translateCurrent(current, layers, parts, options)
+        parts = []
+        translatePath2D(obj, layers, options)
+        break
+
+      default:
+        console.log('##### ERROR')
+        console.log(obj.type)
+        break
+    }
+    // accumlate polygons if necessary
+    if (p instanceof CSG.Polygon) {
```

Comment:
two things : 
- could the different p instanceof XXX not be combined ?
- as with 'instanceof CSG/CAG' perhaps it might be better to test for object properties with some helpers ? (there is always a risk of failure to identify types) : I have a few solutions for that going forward , with changes to CSG.js so this might need to wait ?

Reply:
Since these functions are PRODUCING CSG / CAG / etc objects then the instanceof() methods are valid, i.e. everything being produced is 'correct'.